### ISSUE TITLE
Issue with tsconfig and first file opened

### ISSUE THREAD

**Issue by masad-frost** [typescript-language-server/typescript-language-server#445]
There seems to be an issue where the project's tsconfig isn't properly applied until after the first file is opened, leading to wrong diagnostics and issues for the file. 

Take the following example

Simple directory structure:
- `src/test.tsx`
  - `export const a: string = 1`
- `src/test2.tsx`
  - `export const a: string = 1`
- `tsconfig.json`
  - `{ "compilerOptions": { "jsx": "react-jsx" }, "include": ["./src"] }`

Here is the messages between client and server using basic log level:

<details>
<summary>Click to expand.</summary>

Sent:
```
Content-Length: 1365

{"method":"initialize","params":{"capabilities":{"textDocument":{"synchronization":{"dynamicRegistration":true,"didSave":true},"signatureHelp":{"dynamicRegistration":true,"signatureInformation":{"documentationFormat":["plaintext","markdown"]}},"publishDiagnostics":{"relatedInformation":true},"formatting":{"dynamicRegistration":true},"definition":{"dynamicRegistration":true},"typeDefinition":{"dynamicRegistration":true},"implementation":{"dynamicRegistration":true},"references":{"dynamicRegistration":true},"rename":{"dynamicRegistration":true,"prepareSupport":true},"completion":{"dynamicRegistration":true,"completionItem":{"documentationFormat":["plaintext","markdown"]}},"codeAction":{"dynamicRegistration":true,"codeActionLiteralSupport":{"codeActionKind":{"valueSet":["","quickfix","refactor","refactor.extract","refactor.inline","refactor.rewrite","source","source.organizeImports","source.fixAll"]}}},"hover":{"dynamicRegistration":true,"contentFormat":["plaintext"]}},"workspace":{"applyEdit":true,"configuration":true,"executeCommand":{"dynamicRegistration":true},"workspaceEdit":{"documentChanges":true,"resourceOperations":["create","delete","rename"]}}},"processId":null,"rootUri":"file:///home/runner/lsp-runner","workspaceFolders":[{"name":"file:///home/runner/lsp-runner","uri":"file:///home/runner/lsp-runner"}]},"id":735706350,"jsonrpc":"2.0"}
```

Received:

```
Content-Length: 211

{"jsonrpc":"2.0","method":"window/logMessage","params":{"type":3,"message":"[lspserver] Using Typescript version (workspace) 4.6.4 from path \"/home/runner/lsp-runner/node_modules/typescript/lib/tsserver.js\""}}
```

Received:
```
Content-Length: 1261

{"jsonrpc":"2.0","id":735706350,"result":{"capabilities":{"textDocumentSync":2,"completionProvider":{"triggerCharacters":[".","\"","'","/","@","<"],"resolveProvider":true},"codeActionProvider":{"codeActionKinds":["source.fixAll.ts","source.removeUnused.ts","source.addMissingImports.ts","source.organizeImports.ts"]},"definitionProvider":true,"documentFormattingProvider":true,"documentRangeFormattingProvider":true,"documentHighlightProvider":true,"documentSymbolProvider":true,"executeCommandProvider":{"commands":["_typescript.applyWorkspaceEdit","_typescript.applyCodeAction","_typescript.applyRefactoring","_typescript.organizeImports","_typescript.applyRenameFile"]},"hoverProvider":true,"renameProvider":true,"referencesProvider":true,"signatureHelpProvider":{"triggerCharacters":["(",",","<"]},"workspaceSymbolProvider":true,"implementationProvider":true,"typeDefinitionProvider":true,"foldingRangeProvider":true,"semanticTokensProvider":{"documentSelector":null,"legend":{"tokenTypes":["class","enum","interface","namespace","typeParameter","type","parameter","variable","enumMember","property","function","member"],"tokenModifiers":["declaration","static","async","readonly","defaultLibrary","local"]},"full":true,"range":true},"callsProvider":true}}}
```

Sent:
```
Content-Length: 52

{"method":"initialized","params":{},"jsonrpc":"2.0"}
```

Sent:
```
Content-Length: 206

{"method":"textDocument/didOpen","params":{"textDocument":{"uri":"file:///home/runner/lsp-runner/src/test.tsx","languageId":"javascript","text":"export const a: string = 1;\n","version":0}},"jsonrpc":"2.0"}
```

*Received (problem)*:
```
Content-Length: 331

{"jsonrpc":"2.0","method":"textDocument/publishDiagnostics","params":{"uri":"file:///home/runner/lsp-runner/src/test.tsx","diagnostics":[{"range":{"start":{"line":0,"character":16},"end":{"line":0,"character":22}},"message":"Type annotations can only be used in TypeScript files.","severity":1,"code":8010,"source":"typescript"}]}}
```
</details>

If we wait a bit and send `textDocument/didOpen` for `test2.tsx` (which has the same content as the first file), we get correct diagnostics (i.e. assigning number to string) , not `"Type annotations can only be used in TypeScript files."`. However, if we open the second file quickly after the first file open, both files will have configuration issues.

<details>
<summary>tsserver-log-file</summary>

```
Info 0    [20:18:42.472] Starting TS Server
Info 1    [20:18:42.472] Version: 4.6.4
Info 2    [20:18:42.472] Arguments: /nix/store/ra8r42571xvv1m85wanh1ll9mxmp0mwl-nodejs-16.13.2/bin/node /home/runner/lsp-runner/node_modules/typescript/lib/tsserver.js --logFile ts-logs.txt --logVerbosity verbose --cancellationPipeName /tmp/7bd3c3c98225d0af0e91645a77a06448/tscancellation*
Info 3    [20:18:42.472] Platform: linux NodeVersion: 16 CaseSensitive: true
Info 4    [20:18:42.472] ServerMode: undefined syntaxOnly: false hasUnknownServerMode: undefined
Info 5    [20:18:42.479] Binding...
Info 6    [20:18:42.489] request:
    {"command":"configure","seq":1,"type":"request","arguments":{"formatOptions":{"newLineCharacter":"\n"},"preferences":{"allowIncompleteCompletions":true,"allowRenameOfImportPath":true,"allowTextChangesInNewFiles":true,"displayPartsForJSDoc":true,"generateReturnInDocTemplate":true,"includeAutomaticOptionalChainCompletions":true,"includeCompletionsForImportStatements":true,"includeCompletionsForModuleExports":true,"includeCompletionsWithClassMemberSnippets":true,"includeCompletionsWithInsertText":true,"includeCompletionsWithSnippetText":true,"jsxAttributeCompletionStyle":"auto"}}}
Info 7    [20:18:42.489] Format host information updated
Info 8    [20:18:42.490] response:
    {"seq":0,"type":"response","command":"configure","request_seq":1,"success":true}
Perf 9    [20:18:42.491] 1::configure: async elapsed time (in milliseconds) 1.7859
Info 10   [20:18:42.491] request:
    {"command":"compilerOptionsForInferredProjects","seq":2,"type":"request","arguments":{"options":{"module":"CommonJS","target":"ES2016","jsx":"Preserve","allowJs":true,"allowSyntheticDefaultImports":true,"allowNonTsExtensions":true}}}
Perf 11   [20:18:42.491] 2::compilerOptionsForInferredProjects: elapsed time (in milliseconds) 0.5855
Info 12   [20:18:42.492] response:
    {"seq":0,"type":"response","command":"compilerOptionsForInferredProjects","request_seq":2,"success":true,"body":true}
Info 13   [20:18:42.492] event:
    {"seq":0,"type":"event","event":"typingsInstallerPid","body":{"pid":556}}
Info 14   [20:18:49.440] request:
    {"command":"open","seq":3,"type":"request","arguments":{"file":"/home/runner/lsp-runner/src/test.tsx","fileContent":"export const a: string = 1;\n","scriptKindName":"JS","projectRootPath":"/home/runner/lsp-runner"}}
Info 15   [20:18:49.442] Search path: /home/runner/lsp-runner/src
Info 16   [20:18:49.443] For info: /home/runner/lsp-runner/src/test.tsx :: Config file name: /home/runner/lsp-runner/tsconfig.json
Info 17   [20:18:49.443] Creating configuration project /home/runner/lsp-runner/tsconfig.json
Info 18   [20:18:49.446] FileWatcher:: Added:: WatchInfo: /home/runner/lsp-runner/tsconfig.json 2000 undefined Project: /home/runner/lsp-runner/tsconfig.json WatchType: Config file
Info 19   [20:18:49.448] event:
    {"seq":0,"type":"event","event":"projectLoadingStart","body":{"projectName":"/home/runner/lsp-runner/tsconfig.json","reason":"Creating possible configured project for /home/runner/lsp-runner/src/test.tsx to open"}}
Info 20   [20:18:49.462] Config: /home/runner/lsp-runner/tsconfig.json : {
 "rootNames": [
  "/home/runner/lsp-runner/src/test.tsx",
  "/home/runner/lsp-runner/src/test2.tsx"
 ],
 "options": {
  "jsx": 4,
  "configFilePath": "/home/runner/lsp-runner/tsconfig.json"
 }
}
Info 21   [20:18:49.463] DirectoryWatcher:: Added:: WatchInfo: /home/runner/lsp-runner/src 1 undefined Config: /home/runner/lsp-runner/tsconfig.json WatchType: Wild card directory
Info 22   [20:18:49.464] Elapsed:: 1.1311600003391504ms DirectoryWatcher:: Added:: WatchInfo: /home/runner/lsp-runner/src 1 undefined Config: /home/runner/lsp-runner/tsconfig.json WatchType: Wild card directory
Info 23   [20:18:49.466] FileWatcher:: Added:: WatchInfo: /home/runner/lsp-runner/src/test2.tsx 500 undefined WatchType: Closed Script info
Info 24   [20:18:49.467] Starting updateGraphWorker: Project: /home/runner/lsp-runner/tsconfig.json
Info 25   [20:18:49.482] DirectoryWatcher:: Added:: WatchInfo: /home/runner/lsp-runner/node_modules 1 undefined WatchType: node_modules for closed script infos and package.jsons affecting module specifier cache
Info 26   [20:18:49.514] Elapsed:: 31.673336000181735ms DirectoryWatcher:: Added:: WatchInfo: /home/runner/lsp-runner/node_modules 1 undefined WatchType: node_modules for closed script infos and package.jsons affecting module specifier cache
Info 27   [20:18:50.279] DirectoryWatcher:: Added:: WatchInfo: /home/runner/lsp-runner/src 1 undefined Project: /home/runner/lsp-runner/tsconfig.json WatchType: Failed Lookup Locations
Info 28   [20:18:50.279] Elapsed:: 0.02916999999433756ms DirectoryWatcher:: Added:: WatchInfo: /home/runner/lsp-runner/src 1 undefined Project: /home/runner/lsp-runner/tsconfig.json WatchType: Failed Lookup Locations
Info 29   [20:18:50.280] DirectoryWatcher:: Added:: WatchInfo: /home/runner/lsp-runner/node_modules 1 undefined Project: /home/runner/lsp-runner/tsconfig.json WatchType: Failed Lookup Locations
Info 30   [20:18:50.280] Elapsed:: 0.02508999966084957ms DirectoryWatcher:: Added:: WatchInfo: /home/runner/lsp-runner/node_modules 1 undefined Project: /home/runner/lsp-runner/tsconfig.json WatchType: Failed Lookup Locations
Info 31   [20:18:50.281] DirectoryWatcher:: Added:: WatchInfo: /home/runner/lsp-runner/node_modules/@types 1 undefined Project: /home/runner/lsp-runner/tsconfig.json WatchType: Type roots
Info 32   [20:18:50.282] Elapsed:: 0.6492800004780293ms DirectoryWatcher:: Added:: WatchInfo: /home/runner/lsp-runner/node_modules/@types 1 undefined Project: /home/runner/lsp-runner/tsconfig.json WatchType: Type roots
Info 33   [20:18:50.282] Finishing updateGraphWorker: Project: /home/runner/lsp-runner/tsconfig.json Version: 1 structureChanged: true structureIsReused:: Not Elapsed: 815.1627280004323ms
Info 34   [20:18:50.282] Project '/home/runner/lsp-runner/tsconfig.json' (Configured)
Info 35   [20:18:50.285] 	Files (7)
	/home/runner/lsp-runner/node_modules/typescript/lib/lib.d.ts
	/home/runner/lsp-runner/node_modules/typescript/lib/lib.es5.d.ts
	/home/runner/lsp-runner/node_modules/typescript/lib/lib.dom.d.ts
	/home/runner/lsp-runner/node_modules/typescript/lib/lib.webworker.importscripts.d.ts
	/home/runner/lsp-runner/node_modules/typescript/lib/lib.scripthost.d.ts
	/home/runner/lsp-runner/src/test.tsx
	/home/runner/lsp-runner/src/test2.tsx


	node_modules/typescript/lib/lib.d.ts
	  Default library for target 'es3'
	node_modules/typescript/lib/lib.es5.d.ts
	  Library referenced via 'es5' from file 'node_modules/typescript/lib/lib.d.ts'
	node_modules/typescript/lib/lib.dom.d.ts
	  Library referenced via 'dom' from file 'node_modules/typescript/lib/lib.d.ts'
	node_modules/typescript/lib/lib.webworker.importscripts.d.ts
	  Library referenced via 'webworker.importscripts' from file 'node_modules/typescript/lib/lib.d.ts'
	node_modules/typescript/lib/lib.scripthost.d.ts
	  Library referenced via 'scripthost' from file 'node_modules/typescript/lib/lib.d.ts'
	src/test.tsx
	  Matched by include pattern './src' in 'tsconfig.json'
	src/test2.tsx
	  Matched by include pattern './src' in 'tsconfig.json'

Info 36   [20:18:50.285] -----------------------------------------------
Info 37   [20:18:50.288] FileWatcher:: Added:: WatchInfo: /home/runner/lsp-runner/package.json 250 undefined WatchType: package.json file
Info 38   [20:18:50.295] AutoImportProviderProject: found 1 root files in 1 dependencies in 8.563508999533951 ms
Info 39   [20:18:50.297] Starting updateGraphWorker: Project: /dev/null/autoImportProviderProject1*
Info 40   [20:18:50.682] Finishing updateGraphWorker: Project: /dev/null/autoImportProviderProject1* Version: 1 structureChanged: true structureIsReused:: Not Elapsed: 385.487536999397ms
Info 41   [20:18:50.682] Project '/dev/null/autoImportProviderProject1*' (AutoImportProvider)
Info 42   [20:18:50.682] 	Files (1)
	/home/runner/lsp-runner/node_modules/typescript/lib/typescript.d.ts


	node_modules/typescript/lib/typescript.d.ts
	  Root file specified for compilation

Info 43   [20:18:50.683] -----------------------------------------------
Info 44   [20:18:50.683] event:
    {"seq":0,"type":"event","event":"projectLoadingFinish","body":{"projectName":"/home/runner/lsp-runner/tsconfig.json"}}
Info 45   [20:18:50.685] event:
    {"seq":0,"type":"event","event":"telemetry","body":{"telemetryEventName":"projectInfo","payload":{"projectId":"ce16da1318b3d2e5f209ae1552fab7ab8349f987688f13d828f960ab55ead705","fileStats":{"js":1,"jsSize":28,"jsx":0,"jsxSize":0,"ts":0,"tsSize":0,"tsx":1,"tsxSize":26,"dts":5,"dtsSize":1022068,"deferred":0,"deferredSize":0},"compilerOptions":{"jsx":"react-jsx"},"typeAcquisition":{"enable":false,"include":false,"exclude":false},"extends":false,"files":false,"include":true,"exclude":false,"compileOnSave":false,"configFileName":"tsconfig.json","projectType":"configured","languageServiceEnabled":true,"version":"4.6.4"}}}
Info 46   [20:18:50.687] event:
    {"seq":0,"type":"event","event":"configFileDiag","body":{"triggerFile":"/home/runner/lsp-runner/src/test.tsx","configFile":"/home/runner/lsp-runner/tsconfig.json","diagnostics":[]}}
Info 47   [20:18:50.688] Project '/home/runner/lsp-runner/tsconfig.json' (Configured)
Info 47   [20:18:50.688] 	Files (7)

Info 47   [20:18:50.688] -----------------------------------------------
Info 47   [20:18:50.688] Project '/dev/null/autoImportProviderProject1*' (AutoImportProvider)
Info 47   [20:18:50.688] 	Files (1)

Info 47   [20:18:50.688] -----------------------------------------------
Info 47   [20:18:50.688] Open files: 
Info 47   [20:18:50.688] 	FileName: /home/runner/lsp-runner/src/test.tsx ProjectRootPath: /home/runner/lsp-runner
Info 47   [20:18:50.688] 		Projects: /home/runner/lsp-runner/tsconfig.json
Perf 47   [20:18:50.689] 3::open: async elapsed time (in milliseconds) 1248.2173
Info 48   [20:18:50.689] request:
    {"command":"geterr","seq":4,"type":"request","arguments":{"delay":0,"files":["/home/runner/lsp-runner/src/test.tsx"]}}
Perf 49   [20:18:50.690] 4::geterr: async elapsed time (in milliseconds) 0.5869
Info 50   [20:18:50.713] event:
    {"seq":0,"type":"event","event":"syntaxDiag","body":{"file":"/home/runner/lsp-runner/src/test.tsx","diagnostics":[{"start":{"line":1,"offset":17},"end":{"line":1,"offset":23},"text":"Type annotations can only be used in TypeScript files.","code":8010,"category":"error"}]}}
Info 51   [20:18:50.727] event:
    {"seq":0,"type":"event","event":"semanticDiag","body":{"file":"/home/runner/lsp-runner/src/test.tsx","diagnostics":[]}}
Info 52   [20:18:50.729] event:
    {"seq":0,"type":"event","event":"suggestionDiag","body":{"file":"/home/runner/lsp-runner/src/test.tsx","diagnostics":[]}}
Info 53   [20:18:50.729] event:
    {"seq":0,"type":"event","event":"requestCompleted","body":{"request_seq":4}}
```
</details>

I was able to diagnose this issue using this code https://replit.com/@replitfaris/lsp-runner#index.js, where we spawn the server as a child process and pipe messages to it when you hit enter in the console, feel free to fork and play with it.

This I tried to search for a `typescript`/`typescript-language-server` version combo where this is not an issue. It seems to be reproducible on any version combination of both.

Happy to jump on this personally, if you have any high level pointers.

**Comment by rchl** [comment]
Seems like a client (editor) issue. It's sending `javascript` for `languageId` while it should be `typescriptreact`.

**Comment by masad-frost** [comment]
Nice catch, I guess what threw me off is that it works for consecutive files, it's only the first file that gets it worng.

You can see that here https://replit.com/@replitfaris/lsp-runner#index.js

Here's a recording 

https://user-images.githubusercontent.com/1994028/166588187-c5892567-56df-430d-9cb2-51346fbd24f2.mp4

**Comment by rchl** [comment]
Sure but does it work correctly when passing proper `languageId`?

**Comment by masad-frost** [comment]
Yes thanks
